oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -148,66 +131,227 @@ impl GrpcServerNode {
+        );
+        let result = async_runtime.block_on(server);
+        info!(
+            "{}: Exiting gRPC server pseudo-Node thread {:?}",
+            pretty_name, result
+        );
+
+        self.runtime.exit_node();
     }
+}
 
-    /// Creates a [`GrpcRequest`] instance from a `http_request_path` and an `http_request_body`.
-    fn decode_grpc_request(
-        http_request_path: &str,
-        http_request_body: &dyn hyper::body::Buf,
-    ) -> Result<GrpcRequest, GrpcServerError> {
-        // Parse an HTTP request body as an [`Any`] message.
-        let grpc_request_body = Any::decode(http_request_body.bytes()).map_err(|error| {
-            error!("Failed to parse Protobuf message {}", error);
-            GrpcServerError::BadProtobufMessage
-        })?;
+/// Oak Node implementation for the gRPC server.
+impl super::Node for GrpcServerNode {
+    fn start(self: Box<Self>) -> Result<JoinHandle<()>, OakStatus> {
+        let thread_handle = thread::Builder::new()
+            .name(self.to_string())
+            .spawn(move || self.worker_thread())
+            .expect("Failed to spawn thread");
+        Ok(thread_handle)
+    }
+}
+
+/// [`HttpRequestHandler`] handles HTTP/2 requests from a client and sends HTTP/2 responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to a Runtime that corresponds to a node that created a gRPC server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing gRPC invocations.
+    writer: Handle,
+}
+
+// Set a mandatory prefix for all gRPC requests processed by a gRPC pseudo-Node.
+impl NamedService for HttpRequestHandler {
+    const NAME: &'static str = "oak";
+}
+
+impl Service<http::Request<hyper::Body>> for HttpRequestHandler {
+    type Response = http::Response<tonic::body::BoxBody>;
+    type Error = http::Error;
+    type Future = BoxFuture<Self::Response, Self::Error>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    // Decodes an unary gRPC request using a [`VecCodec`] and processes it with
+    // [`tonic::server::Grpc::unary`] and a [`GrpcRequestHandler`].
+    fn call(&mut self, request: http::Request<hyper::Body>) -> Self::Future {
+        let grpc_handler = GrpcRequestHandler::new(
+            self.runtime.clone(),
+            self.writer,
+            request.uri().path().to_string(),
+        );
+
+        let future = async move {
+            debug!("Processing an HTTP/2 request: {:?}", request);
+            let mut grpc_service = Grpc::new(VecCodec::default());
+            let response = grpc_service.unary(grpc_handler, request).await;
+            debug!("Sending an HTTP/2 response: {:?}", response);
+            Ok(response)
+        };
+
+        Box::pin(future)
+    }
+}
+
+/// Encapsulates [`VecEncoder`] and [`VecDecoder`] types and is used by [`tonic::server::Grpc`].
+#[derive(Clone, Debug, Default)]
```

Comment:
Does this really need to be `Clone`, `Debug` and `Default`? Are these needed by `tonic` itself? (and below)

Reply:
Deleted, only `Default` is left